scipy/stats/_continuous_distns.py

```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+                out[idx] = self._rvs_scalar(it[0], it[1],
+                                            numsamples).reshape(shp)
+                it.iternext()
+
+        if self._size == ():
+            out = out[()]
+        return out
+
+    def _rvs_scalar(self, p, b, numsamples=None):
+        invert_res = False
+        sz, rndm = numsamples, self._random_state
+        if not(sz):
+            sz = 1
+        if p < 0:
+            # note: if X is geninvgauss(p, b), then 1/X is geninvgauss(-p, b)
+            p = -p
+            invert_res = True
+        m = self._mode(p, b)
+        if (p >= 1) | (b > 1):
+            # ratio of uniforms with mode shift
+            a2 = -2 * (p + 1) / b - m
+            a1 = 2 * m * (p - 1) / b - 1
+            # find roots of x**3 + a2 * x**2 + a1 * x + m (Cardano's formula)
+            p1 = a1 - a2**2 / 3
+            q1 = 2 * a2**3 / 27 - a2 * a1 / 3 + m
+            phi = np.arccos(-q1 * np.sqrt(-27 / p1**3) / 2)
+            s1 = -np.sqrt(-4 * p1 / 3)
+            root1 = s1 * np.cos(phi / 3 + np.pi / 3) - a2 / 3
+            root2 = -s1 * np.cos(phi / 3) - a2 / 3
+            # root3 = s1 * np.cos(phi / 3 - np.pi / 3) - a2 / 3
+            vmin = (root1 - m) * np.sqrt(self.pdf(root1, p, b))
+            vmax = (root2 - m) * np.sqrt(self.pdf(root2, p, b))
+            umax = np.sqrt(self._pdf(m, p, b))
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=vmin, vmax=vmax, c=m,
+                                     random_state=rndm)
+        elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
+            # ratio of uniforms without mode shift
+            umax = np.sqrt(self._pdf(m, p, b))
+            s = sc.kv(p + 2, b) / sc.kv(p, b)
+            vmax = np.sqrt(self._pdf(self._mode(p + 2, b), p + 2, b) * s)
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=0, vmax=vmax, c=0,
+                                     random_state=rndm)
+        else:
+            # new algorithm in [2]
+            x0 = b / (1 - p)
+            xs = np.max((x0, 2 / b))
+            k1 = self._pdf(m, p, b)
+            A1 = k1 * x0
+            if x0 < 2 / b:
+                k2 = np.exp(-b)
+                if p > 0:
+                    A2 = k2 * ((2 / b)**p - x0**p) / p
+                else:
+                    A2 = k2 * np.log(2 / b**2)
+            else:
+                k2, A2 = 0, 0
+            k3 = xs**(p - 1)
+            A3 = 2 * k3 * np.exp(-xs * b / 2) / b
+            A = A1 + A2 + A3
+
+            size1d = tuple(np.atleast_1d(sz))
+            N = np.prod(size1d)  # number of rvs needed, reshape upon return
+
+            # start sampling
+            rng = self._random_state
+            x = np.zeros(N)
+            simulated = 0
+            # [2]: rejection constant is < 2.73; so expected runtime is finite
+            while True:
```

Comment:
This loop begins with `while True:`, and is terminated with a `break` in an `if` statement:

                if simulated >= N:
                    rvs = np.reshape(x, size1d)
                    break

You have constructed the calculation so that `simulated` *can't* exceed `N`, and the loop only terminates when `simulated == N`.  So it would be cleaner to start the loop with

            while simulated < N:

then remove the `if` statement that checks for the termination condition, and after the while loop add the line

            rvs = np.reshape(x, size1d)

Reply:
you are right, fixed